### PR TITLE
Introduce Realm Object Server API

### LINKED ISSUE #860: ROS api

### PR DESCRIPTION
**Pull request by kristiandupont**
This resolves #860.

### CODE REVIEW

File: Realm.Shared/User.cs [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+        {
+            throw new NotImplementedException();
+        }
+
+        public static User FromJson()
+        {
+            throw new NotImplementedException();
+        }
+
+        public static User Authenticate(Credentials credentials, string serverUrl)
```
Comment (fealebenpae) [inline Realm.Shared/User.cs]: Unless we really need to have it, I'd suggest nixing the synchronous API.

File: Realm.Shared/User.cs [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+namespace Realms
+{
+    public class User
+    {
+        public static User GetCurrentUser()
+        {
+            throw new NotImplementedException();
+        }
+
+        public static User FromJson()
```
Comment (fealebenpae) [inline Realm.Shared/User.cs]: I would suggest we have `User` implement `ISerializable` which Json.NET and other serializers support natively.

File: Realm.Shared/User.cs [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+        public string Identity { get; private set; }
+
+        public bool IsValid { get; private set; }
+
+        public void LogOut()
+        {
+            throw new NotImplementedException();
+        }
+
+        public string ToJson()
```
Comment (fealebenpae) [inline Realm.Shared/User.cs]: Same as the comment I left on `User.FromJson`.

File: Realm.Shared/Credentials.cs [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+
+using System;
+using System.Collections.Generic;
+
+namespace Realms
+{
+    public class Credentials
+    {
+        public static Credentials Custom(string identityProvider, string userIdentifier,
+            Dictionary<string, object> userInfo)
```
Comment (fealebenpae) [inline Realm.Shared/Credentials.cs]: I would suggest making this an `IDictionary<string, object>`.

File: Realm.Shared/Session.cs [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+        Initial,
+        Stopped,
+        Unbound
+    }
+
+    public class Session
+    {
+        public SyncConfiguration Configuration { get; private set; }
+        public Uri ServerUri { get; private set; }
+        public SessionState State { get; private set; }
```
Comment (fealebenpae) [inline Realm.Shared/Session.cs]: Can we expose a `StateChanged` event or something similar? I mean - do we get notifications like that from Object Store?

File: Realm.Shared/User.cs [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+////////////////////////////////////////////////////////////////////////////
+
+using System;
+using System.Threading.Tasks;
+
+namespace Realms
+{
+    public class User
+    {
+        public static User GetCurrentUser()
```
Comment (fealebenpae) [inline Realm.Shared/User.cs]: This might work better as a property. Or field.

File: Realm.Shared/User.cs [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+            throw new NotImplementedException();
+        }
+
+        public string AccessToken { get; private set; }
+
+        public string Identity { get; private set; }
+
+        public bool IsValid { get; private set; }
+
+        public void LogOut()
```
Comment (fealebenpae) [inline Realm.Shared/User.cs]: If the implementation allows it, this method should return a `Task` as well.

File: Realm.Shared/User.cs (hunk 6 above)
Comment (nirinchev) [inline Realm.Shared/User.cs]: Property! Fields are evil!

File: Platform.XamarinAndroid/Realm.Sync.XamarinAndroid/Properties/AssemblyInfo.cs [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+////////////////////////////////////////////////////////////////////////////
+
+using System.Reflection;
+using System.Runtime.CompilerServices;
+
+[assembly: AssemblyTitle("Realm.Sync.XamarinAndroid")]
+[assembly: InternalsVisibleTo("Tests.XamarinAndroid")]
```
Comment (nirinchev) [inline Platform.XamarinAndroid/Realm.Sync.XamarinAndroid/Properties/AssemblyInfo.cs]: Is there LinkerSafe on Android?

File: Platform.XamarinIOS/Realm.Sync.XamarinIOS/ApiDefinition.cs [hunk 9]
```diff
@@ -0,0 +1,62 @@
+﻿// WARNING even though this file contains apparently generic content with no function
```
Comment (nirinchev) [inline Platform.XamarinIOS/Realm.Sync.XamarinIOS/ApiDefinition.cs]: Seems like the header is missing

File: Shared/Realm.Sync.Shared/SyncConfiguration.cs [hunk 10]
```diff
@@ -17,6 +17,7 @@
 ////////////////////////////////////////////////////////////////////////////
 
 using System;
+using System.Runtime.CompilerServices;
```
Comment (nirinchev) [inline Shared/Realm.Sync.Shared/SyncConfiguration.cs]: why is this needed?

**Comment by fealebenpae** [comment]
retest this please

File: Shared/Realm.Shared/RealmConfiguration.cs [hunk 11]
```diff
@@ -239,6 +235,56 @@ public bool Equals(RealmConfiguration rhs)
...
+        internal virtual void DeleteRealm()
+        {
+            // TODO add cache checking when implemented, https://github.com/realm/realm-dotnet/issues/308
+            // when cache checking, uncomment in IntegrationTests.cs RealmInstanceTests.DeleteRealmFailsIfOpenSameThread and add a variant to test open on different thread
+            var lockOnWhileDeleting = new object();
+            lock (lockOnWhileDeleting)
+            {
+                var fullpath = DatabasePath;
+                File.Delete(fullpath);
+                File.Delete(fullpath + ".log_a");  // eg: name at end of path is EnterTheMagic.realm.log_a
```
Comment (AndyDentFree) [inline Shared/Realm.Shared/RealmConfiguration.cs]: Is this set of files still accurate in current core? I thought the log had been moved into the Realm?

File: Shared/Realm.Sync.Shared/Credentials.cs [hunk 12]
```diff
@@ -0,0 +1,162 @@
...
+using System.Collections.Generic;
+using System.Json;
+
+namespace Realms.Sync
+{
+    public class Credentials
+    {
+        internal static class Providers
+        {
+            internal const string Debug = "debug";
```
Comment (AndyDentFree) [inline Shared/Realm.Sync.Shared/Credentials.cs]: Are these strings going to get moved into ObjectStore?

File: Shared/Realm.Sync.Shared/Native/SyncConfiguration.cs [hunk 13]
```diff
@@ -0,0 +1,52 @@
...
+////////////////////////////////////////////////////////////////////////////
+
+using System;
+using System.Diagnostics.CodeAnalysis;
+using System.Runtime.InteropServices;
+
+namespace Realms.Sync.Native
+{
+    [StructLayout(LayoutKind.Sequential)]
+    [SuppressMessage("StyleCop.CSharp.NamingRules", "SA1307:AccessibleFieldsMustBeginWithUpperCaseLetter")]
```
Comment (kristiandupont) [inline Shared/Realm.Sync.Shared/Native/SyncConfiguration.cs]: Is this suppression still necessary? There doesn't seem to be any accessible fields?

File: Shared/Realm.Sync.Shared/RealmExtensions.cs [hunk 14]
```diff
@@ -0,0 +1,30 @@
...
+//
+////////////////////////////////////////////////////////////////////////////
+
+using System;
+
+namespace Realms.Sync
+{
+    public static class RealmExtensions
+    {
+        public static Session GetSession(this Realm @this)
```
Comment (kristiandupont) [inline Shared/Realm.Sync.Shared/RealmExtensions.cs]: We could remove all things *session* from this PR if we wanted.

File: Shared/Realm.Sync.Shared/Credentials.cs (hunk 12 above)
Comment (fealebenpae) [inline Shared/Realm.Sync.Shared/Credentials.cs]: Not sure - they're only used for REST calls to the object server and that's unlikely to make it in object store.

File: Shared/Realm.Shared/RealmConfiguration.cs (hunk 11 above)
Comment (fealebenpae) [inline Shared/Realm.Shared/RealmConfiguration.cs]: It might be stale - @kristiandupont suggested moving this functionality to object store going forward might be better in the long run.

File: Shared/Tests.Sync.Shared/SyncConfigurationTests.cs [hunk 15]
```diff
@@ -0,0 +1,51 @@
...
+        }
+
+        [TearDown]
+        public void TearDown()
+        {
+            _realm.Dispose();
+        }
+
+        [Test]
+        public void Foo()
```
Comment (kristiandupont) [inline Shared/Tests.Sync.Shared/SyncConfigurationTests.cs]: Hehe well I guess we can safely remove this test...

File: Platform.XamarinIOS/Realm.Sync.XamarinIOS/ApiDefinition.cs (hunk 9 above)
Comment (fealebenpae) [inline Platform.XamarinIOS/Realm.Sync.XamarinIOS/ApiDefinition.cs]: It's the same as in `Realm.XamarinIOS`. Since it basically comes from the Xamarin Studio template, I don't know if we need or want to have the license header :/

File: Platform.XamarinAndroid/Realm.Sync.XamarinAndroid/Properties/AssemblyInfo.cs (hunk 8 above)
Comment (fealebenpae) [inline Platform.XamarinAndroid/Realm.Sync.XamarinAndroid/Properties/AssemblyInfo.cs]: Not that I could find.

File: Shared/Realm.Sync.Shared/RealmExtensions.cs (hunk 14 above)
Comment (fealebenpae) [inline Shared/Realm.Sync.Shared/RealmExtensions.cs]: Not all, we still need the `SessionHandle` for authentication :P

I could remove this, but I'd only end up adding it again later :/

File: Shared/Tests.Sync.Shared/SyncConfigurationTests.cs (hunk 15 above)
Comment (fealebenpae) [inline Shared/Tests.Sync.Shared/SyncConfigurationTests.cs]: Haha, good catch!

File: Shared/Realm.Sync.Shared/RealmExtensions.cs (hunk 14 above)
Comment (kristiandupont) [inline Shared/Realm.Sync.Shared/RealmExtensions.cs]: Yeah, let's just keep it then.

Comment (AndyDentFree) [inline Shared/Realm.Sync.Shared/RealmExtensions.cs]: What is this for? missing comment for API doc.

File: Shared/Realm.Sync.Shared/SyncConfiguration.cs [hunk 16]
```diff
@@ -0,0 +1,55 @@
...
+// limitations under the License.
+//
+////////////////////////////////////////////////////////////////////////////
+
+using System;
+using System.Runtime.CompilerServices;
+
+namespace Realms.Sync
+{
+    public class SyncConfiguration : RealmConfiguration
```
Comment (AndyDentFree) [inline Shared/Realm.Sync.Shared/SyncConfiguration.cs]: Needs XML doc.

Comment (fealebenpae) [inline Shared/Realm.Sync.Shared/SyncConfiguration.cs]: I've added an issue to track API docs - https://github.com/realm/realm-dotnet/issues/929.

My hope is to get this merged soon so that we can tackle the issues, including documentation, in parallel.

**Comment by AndyDentFree** [comment]
Lets make documenting the new stuff a separate issue but must be 1.0!

**Comment by AndyDentFree** [comment]
👍

File: Shared/Realm.Shared/RealmConfiguration.cs [hunk 17]
```diff
@@ -239,6 +235,56 @@ public bool Equals(RealmConfiguration rhs)
...
+
+            srHandle.SetHandle(srPtr);
+            return new Realm(srHandle, this, schema);
+        }
+
+        internal virtual void DeleteRealm()
+        {
+            // TODO add cache checking when implemented, https://github.com/realm/realm-dotnet/issues/308
+            // when cache checking, uncomment in IntegrationTests.cs RealmInstanceTests.DeleteRealmFailsIfOpenSameThread and add a variant to test open on different thread
+            var lockOnWhileDeleting = new object();
```
Comment (nirinchev) [inline Shared/Realm.Shared/RealmConfiguration.cs]: What is this lock doing?

File: Shared/Realm.Sync.Shared/Credentials.cs [hunk 18]
```diff
@@ -0,0 +1,162 @@
...
+
+            internal const string Password = "password";
+
+            internal const string Identity = "identity";
+
+            internal const string IsAdmin = "isAdmin";
+        }
+
+        public static Credentials Custom(string identityProvider, string userIdentifier,
+            IReadOnlyDictionary<string, object> userInfo)
```
Comment (nirinchev) [inline Shared/Realm.Sync.Shared/Credentials.cs]: Do we want to force `userInfo` to be passed as readonly?

File: Shared/Realm.Sync.Shared/Credentials.cs [hunk 19]
```diff
@@ -0,0 +1,162 @@
...
+
+        public IReadOnlyDictionary<string, object> UserInfo { get; private set; } = new Dictionary<string, object>();
+
+        private Credentials()
+        {
+        }
+
+        internal JsonObject ToJson()
+        {
+            var user_info = new JsonObject();
```
Comment (nirinchev) [inline Shared/Realm.Sync.Shared/Credentials.cs]: That's not c++ ;)

File: Shared/Realm.Sync.Shared/SyncConfiguration.cs [hunk 20]
```diff
@@ -0,0 +1,55 @@
...
+    {
+        public Uri ServerUri { get; private set; }
+
+        public User User { get; private set; }
+
+        public bool ShouldDeleteRealmOnLogOut { get; private set; }
+
+        public SyncConfiguration(User user, Uri serverUri)
```
Comment (nirinchev) [inline Shared/Realm.Sync.Shared/SyncConfiguration.cs]: Can we use string instead of an `Uri` here? I've always felt `Uri`-based API are somewhat cumbersome. We could always do a `var uri = new Uri(ServerUri)` whenever we need.

File: Shared/Realm.Sync.Shared/SyncConfiguration.cs [hunk 21]
```diff
@@ -0,0 +1,55 @@
...
+        {
+            this.User = user;
```
Comment (nirinchev) [inline Shared/Realm.Sync.Shared/SyncConfiguration.cs]: Are we doing `this`-es now? 😱 🙏

File: Shared/Realm.Sync.Shared/User.cs [hunk 22]
```diff
@@ -0,0 +1,152 @@
...
+        {
+            var json = new JsonObject
+            {
+                ["data"] = RefreshToken,
+                ["path"] = path,
+                ["provider"] = "realm"
+            };
+
+            var result = await MakeAuthRequestAsync(ServerUri, json, TimeSpan.FromSeconds(30)).ConfigureAwait(continueOnCapturedContext: false);
+            var access_token = result["access_token"];
```
Comment (nirinchev) [inline Shared/Realm.Sync.Shared/User.cs]: The c# variable doesn't have to be named like the json field

File: wrappers/src/sync_manager_cs.cpp [hunk 23]
```diff
@@ -0,0 +1,86 @@
...
+    std::shared_ptr<SyncUser>* user;
+
+    uint16_t* url;
+    size_t url_len;
+};
+
+void (*s_refresh_access_token_callback)(std::shared_ptr<SyncUser>*, std::shared_ptr<SyncSession>*, const char* path, size_t path_len);
+
+extern "C" {
+REALM_EXPORT void realm_initialize_sync(const uint16_t* base_path_buf, size_t base_path_len, decltype(s_refresh_access_token_callback) refresh_callback)
```
Comment (nirinchev) [inline wrappers/src/sync_manager_cs.cpp]: Should the other arguments be const as well?

File: wrappers/src/sync_manager_cs.cpp [hunk 24]
```diff
@@ -0,0 +1,86 @@
...
+}
+    
+    
+REALM_EXPORT SharedRealm* shared_realm_open_with_sync(Configuration configuration, SyncConfiguration sync_configuration, SchemaObject* objects, int objects_length, SchemaProperty* properties, uint8_t* encryption_key, NativeException::Marshallable& ex)
+{
+    return handle_errors(ex, [&]() {
+        Realm::Config config;
+        config.schema_mode = SchemaMode::Additive;
+
+        // by definition the key is only allowwed to be 64 bytes long, enforced by C# code
```
Comment (nirinchev) [inline wrappers/src/sync_manager_cs.cpp]: Typo - `allowwed`

File: wrappers/src/sync_manager_cs.cpp [hunk 25]
```diff
@@ -0,0 +1,86 @@
...
+        if (encryption_key)
+          config.encryption_key = std::vector<char>(encryption_key, encryption_key+64);
```
Comment (nirinchev) [inline wrappers/src/sync_manager_cs.cpp]: Missing curly braces + wrong indentation (2 instead of 4 spaces)

File: Shared/Realm.Sync.Shared/Credentials.cs (hunk 19 above)
Comment (fealebenpae) [inline Shared/Realm.Sync.Shared/Credentials.cs]: It's not, but it's the name for the field in the JSON dictionary in the authentication protocol.

File: wrappers/src/sync_manager_cs.cpp (hunk 23 above)
Comment (fealebenpae) [inline wrappers/src/sync_manager_cs.cpp]: They're primitives - we can't change their contents the same way as the string buffer so there's no need to denote them as `const`.

File: Shared/Realm.Sync.Shared/Credentials.cs (hunk 18 above)
Comment (fealebenpae) [inline Shared/Realm.Sync.Shared/Credentials.cs]: You can still pass a `Dictionary<string, object>` because it, strangely enough, implements the interface.

File: Shared/Realm.Sync.Shared/SyncConfiguration.cs (hunk 20 above)
Comment (fealebenpae) [inline Shared/Realm.Sync.Shared/SyncConfiguration.cs]: I prefer using `Uri` because this puts the onus on the user of the API to make sure they have a well-formed URI string to begin with.

Comment (fealebenpae) [inline Shared/Realm.Sync.Shared/SyncConfiguration.cs]: Also, Cocoa uses `NSURL` here.

File: Shared/Realm.Shared/RealmConfiguration.cs (hunk 17 above)
Comment (fealebenpae) [inline Shared/Realm.Shared/RealmConfiguration.cs]: Sigh, this was just moved here from somewhere else. I'll just undo the refactoring of the deletion logic.

File: Shared/Realm.Sync.Shared/User.cs (hunk 22 above)
Comment (fealebenpae) [inline Shared/Realm.Sync.Shared/User.cs]: You're right, but I find using the same name helps me reason about things better. We also use snake_case for identifiers that denote something from our C++ layer like `DllImport`-ed functions or marshaled structures, so I could argue there is precedent for this.

File: Shared/Realm.Sync.Shared/SyncConfiguration.cs (hunk 20 above)
Comment (nirinchev) [inline Shared/Realm.Sync.Shared/SyncConfiguration.cs]: That's what I dislike ;) In almost all cases when an API requires `Uri`, I'll pass in `new Uri(myStringUri)`, which feels annoying and cumbersome. We can always do an `Uri.TryParse` and if it's not well formed throw an exception to let the user know something's wrong.

File: Shared/Realm.Sync.Shared/SyncConfiguration.cs (hunk 21 above)
Comment (nirinchev) [inline Shared/Realm.Sync.Shared/SyncConfiguration.cs]: I don't think there's a rule for that. There is one for forcing `this`-es (that's disabled), but it doesn't come with one that forces their absence.

File: wrappers/src/sync_manager_cs.cpp (hunk 25 above)
Comment (nirinchev) [inline wrappers/src/sync_manager_cs.cpp]: But then you would never learn! Which is the point of those reviews 😜

File: Shared/Realm.Sync.Shared/Credentials.cs (hunk 18 above)
Comment (fealebenpae) [inline Shared/Realm.Sync.Shared/Credentials.cs]: Yeah, you're right - that would be better. I'll change it.

File: Shared/Realm.Sync.Shared/SyncConfiguration.cs (hunk 20 above)
Comment (fealebenpae) [inline Shared/Realm.Sync.Shared/SyncConfiguration.cs]: @AndyDentFree @kristiandupont: `string` vs `Uri`?

File: Shared/Realm.Shared/Realm.cs [hunk 26]
```diff
@@ -301,7 +301,19 @@ public override int GetHashCode()
         /// <param name="configuration">A configuration which supplies the realm path.</param>
         public static void DeleteRealm(RealmConfiguration configuration)
         {
-            configuration.DeleteRealm();
+            // TODO add cache checking when implemented, https://github.com/realm/realm-dotnet/issues/308
+            // when cache checking, uncomment in IntegrationTests.cs RealmInstanceTests.DeleteRealmFailsIfOpenSameThread and add a variant to test open on different thread
+            var lockOnWhileDeleting = new object();
+            lock (lockOnWhileDeleting)
```
Comment (nirinchev) [inline Shared/Realm.Shared/Realm.cs]: I know this is just moving things to where they were, but this lock seems wrong, do you feel like refactoring it or adding a TODO to fix it later? :)

File: Shared/Realm.Sync.Shared/SyncConfiguration.cs (hunk 20 above)
Comment (kristiandupont) [inline Shared/Realm.Sync.Shared/SyncConfiguration.cs]: I am not strongly in one camp or the other. I do think it's nice to emphasize that this is not the path. That would have been a bigger problem if the user parameter wasn't there, though..
